### ISSUE TITLE
pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5

### ISSUE THREAD

**Issue by wandaoyi** [ultralytics/yolov5#1585]
However, I want to know about why:
pxy = ps[:, :2].sigmoid() * 2. - 0.5
pwh = (ps[:, 2:4].sigmoid() * 2) ** 2 * anchors[i]

do    * 2. - 0.5 and * 2) ** 2    these things.
I don't understand. so, I wonder!
## ❔Question

        pxy = ps[:, :2].sigmoid() * 2. - 0.5
        pwh = (ps[:, 2:4].sigmoid() * 2) ** 2 * anchors[i]
        
        pbox = torch.cat((pxy, pwh), 1).to(device)  # predicted box
        #print(pbox)
        giou = bbox_iou(pbox.T, tbox[i], x1y1x2y2=False, CIoU=True)  # giou(prediction, target)
        #print(tbox[i])
        print(len(giou))
        lbox += (1.0 - giou).mean()  # giou loss

**Comment by zhiqwang** [comment]
In general, object detection network doesn't learn the bbox coordinates directly, the targets for them to learn are the correction offset with predefined *anchor boxes*.

**Comment by wandaoyi** [comment]
Thank you. And what's the math's meaning? could you talk about it?




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Zhiqiang Wang"<notifications@github.com&gt;; 
发送时间: 2020年12月3日(星期四) 下午3:16
收件人: "ultralytics/yolov5"<yolov5@noreply.github.com&gt;; 
抄送: "求道"<1972241575@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [ultralytics/yolov5] pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5 (#1585)





 
In general, object detection network doesn't learn the bbox coordinates directly, the targets for them to learn are the correction offset with the Anchor.

**Comment by zhiqwang** [comment]
This #471 is a good reference.

**Comment by wandaoyi** [comment]
Thank you very much!&nbsp;


Ori formula: b_x = sigmoid(t_x) * 2 - 0.5 + c_x


But, in the yolov5:&nbsp;pxy = ps[:, :2].sigmoid() * 2. - 0.5  It didn't add c_x, x_y






------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Zhiqiang Wang"<notifications@github.com&gt;; 
发送时间: 2020年12月3日(星期四) 下午3:40
收件人: "ultralytics/yolov5"<yolov5@noreply.github.com&gt;; 
抄送: "求道"<1972241575@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [ultralytics/yolov5] pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5 (#1585)





 
This #471 is a good reference.

**Comment by zhiqwang** [comment]
Hi @wandaoyi , your typing showing in GitHub is a little confusing, and I guess that `self.grid`  as below should be understand to `c_x` and `c_y`

https://github.com/ultralytics/yolov5/blob/f0101475788590720a3a4b1152a89d531f311dff/models/yolo.py#L57-L58

**Comment by wandaoyi** [comment]
In&nbsp;https://github.com/ultralytics/yolov5/issues/471 : &nbsp;The function under the "And try to formularize it:" word in the page

The function is :&nbsp;
b_x = sigmoid(t_x) * 2 - 0.5 + c_x
b_y = sigmoid(t_y) * 2 - 0.5 + c_y


But in the code:pxy = ps[:, :2].sigmoid() * 2. - 0.5
pwh = (ps[:, 2:4].sigmoid() * 2) ** 2 * anchors[i]




It means:
b_x = sigmoid(t_x) * 2 -  0.5
b_y = sigmoid(t_y) * 2 - 0.5


So, where is our c_x and c_y ?







------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Zhiqiang Wang"<notifications@github.com&gt;; 
发送时间: 2020年12月3日(星期四) 晚上6:26
收件人: "ultralytics/yolov5"<yolov5@noreply.github.com&gt;; 
抄送: "求道"<1972241575@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [ultralytics/yolov5] pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5 (#1585)





 
Hi @wandaoyi , your typing showing here is a little confusing, and I guess that self.grid should be understand as c_x and c_y
 
https://github.com/ultralytics/yolov5/blob/f0101475788590720a3a4b1152a89d531f311dff/models/yolo.py#L57-L58

**Comment by zhiqwang** [comment]
Your formula miss the `self.grid[i]` as in L57, it play a role of `c_x` and `c_y`

**Comment by wandaoyi** [comment]
I don't understand the code in:
yolov5/utils/utils.py
Lines 472 to 475 in 1e95337



# GIoU&nbsp;
pxy = ps[:, :2].sigmoid() * 2. - 0.5&nbsp;
pwh = (ps[:, 2:4].sigmoid() * 2) ** 2 * anchors[i]&nbsp;
pbox = torch.cat((pxy, pwh), 1).to(device) &nbsp;# predicted box&nbsp;



There are not having any grid infos
So, there is nothing to role of c_x and c_y




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Zhiqiang Wang"<notifications@github.com&gt;; 
发送时间: 2020年12月3日(星期四) 晚上7:14
收件人: "ultralytics/yolov5"<yolov5@noreply.github.com&gt;; 
抄送: "求道"<1972241575@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [ultralytics/yolov5] pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5 (#1585)





 
Your formula miss the self.grid[i] as in L57, it play a role of c_x and c_y

**Comment by zhiqwang** [comment]
Hi @wandaoyi , So sad a story, I can't find the link you point to, I think you should double check the typing and the link showing on the GitHub pages.

**Comment by wandaoyi** [comment]
OMG, The code of that in /utils/loss.py function: compute_loss (120~121 line) at yolov5 version of 3.0&nbsp;But at 1.0 and 2.0 version, it's in&nbsp;utils/utils.py&nbsp;function: compute_loss (1.0: 452~453; 2.0: 474~475)




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Zhiqiang Wang"<notifications@github.com&gt;; 
发送时间: 2020年12月3日(星期四) 晚上7:24
收件人: "ultralytics/yolov5"<yolov5@noreply.github.com&gt;; 
抄送: "求道"<1972241575@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [ultralytics/yolov5] pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5 (#1585)





 
Hi @wandaoyi , So sad a story, I can't find the link you point to, I think you should double check the typing and the link showing on the GitHub pages.

**Comment by zhiqwang** [comment]
Ah-hah, The learning stage don't need to add `c_x` and `c_y`, the formula concluded as in #471 is just used at the inference stage.

**Comment by zhiqwang** [comment]
> In general, object detection network doesn't learn the bbox coordinates directly, the targets for them to learn are the correction offset with predefined anchor boxes.

And that is the meaning of this line, we actually just learn the **offset** with predefined anchor boxes.

**Comment by wandaoyi** [comment]
Thank you very much! Got it




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Zhiqiang Wang"<notifications@github.com&gt;; 
发送时间: 2020年12月3日(星期四) 晚上7:50
收件人: "ultralytics/yolov5"<yolov5@noreply.github.com&gt;; 
抄送: "求道"<1972241575@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [ultralytics/yolov5] pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5 (#1585)





  
In general, object detection network doesn't learn the bbox coordinates directly, the targets for them to learn are the correction offset with predefined anchor boxes.
  
And that is the meaning of this line, we actually just learn the offset with predefined anchor boxes.

**Comment by egbit** [comment]
And if you're trying to actually understand the `n * 2 - 0.5` part...  At first it looked to me like a coding error, because converting from a range of 0..1 to a range of -1..1 requires `(n - 0.5) * 2` which is similar.  It's not an error.  What's actually going on is the image is subdivided into a grid of squares, and the coordinates in `grid[]` are the coordinates of the upper-left corner of that square.  The neural network provides x,y coordinates in the range 0..1 (enforced by sigmoid) which covers the square, centered at 0.5.  Multiplying by two allows detected x,y coordinates to cover a larger range, slightly outside the square -- otherwise it's difficult to detect objects centered at grid boundaries.  Subtracting 0.5 shifts the resulting range to -0.5 .. 1.5 which is centered around 0 .. 1.

**Comment by wandaoyi** [comment]
Thank you very much.
I saw it in issue about bbox function:
b_x = sigmoid(t_x) * 2 - 0.5 + c_x
b_y = sigmoid(t_y) * 2 - 0.5 + c_y
In compute_loss(), there are no c_x or c_y, due to training is not needing about &nbsp;that




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "egbit"<notifications@github.com&gt;; 
发送时间: 2020年12月5日(星期六) 上午6:39
收件人: "ultralytics/yolov5"<yolov5@noreply.github.com&gt;; 
抄送: "求道"<1972241575@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [ultralytics/yolov5] pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5 (#1585)





 
And if you're trying to actually understand the n * 2 - 0.5 part...  At first it looked to me like a coding error, because converting from a range of 0..1 to a range of -1..1 requires (n - 0.5) * 2 which is similar.  It's not an error.  What's actually going on is the image is subdivided into a grid of squares, and the coordinates in grid[] are the coordinates of the upper-left corner of that square.  The neural network provides x,y coordinates in the range 0..1 (enforced by sigmoid) which covers the square, centered at 0.5.  Multiplying by two allows detected x,y coordinates to cover a larger range, slightly outside the square -- otherwise it's difficult to detect objects centered at grid boundaries.  Subtracting 0.5 shifts the resulting range to -0.5 .. 1.5 which is centered around 0 .. 1.

**Comment by xuanzhangyang** [comment]
> Thank you very much. I saw it in issue about bbox function: b_x = sigmoid(t_x) * 2 - 0.5 + c_x b_y = sigmoid(t_y) * 2 - 0.5 + c_y In compute_loss(), there are no c_x or c_y, due to training is not needing about &nbsp;that
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "egbit"<notifications@github.com&gt;; 发送时间: 2020年12月5日(星期六) 上午6:39 收件人: "ultralytics/yolov5"<yolov5@noreply.github.com&gt;; 抄送: "求道"<1972241575@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 主题: Re: [ultralytics/yolov5] pxy = ps[:, :2].sigmoid() * 2. - 0.5; why the sigmoid number multiply by 2 and minus 0.5 (#1585) And if you're trying to actually understand the n * 2 - 0.5 part... At first it looked to me like a coding error, because converting from a range of 0..1 to a range of -1..1 requires (n - 0.5) * 2 which is similar. It's not an error. What's actually going on is the image is subdivided into a grid of squares, and the coordinates in grid[] are the coordinates of the upper-left corner of that square. The neural network provides x,y coordinates in the range 0..1 (enforced by sigmoid) which covers the square, centered at 0.5. Multiplying by two allows detected x,y coordinates to cover a larger range, slightly outside the square -- otherwise it's difficult to detect objects centered at grid boundaries. Subtracting 0.5 shifts the resulting range to -0.5 .. 1.5 which is centered around 0 .. 1. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

in utils/loss.py build_targets function， tbox.append(torch.cat((gxy - gij, gwh), 1)) , note this: gxy - gij,  when build target, c_x and c_y has been subtracted。
pxy = ps[:, :2].sigmoid() * 2. - 0.5
pwh = (ps[:, 2:4].sigmoid() * 2) ** 2 * anchors[i]
pbox = torch.cat((pxy, pwh), 1)
iou = bbox_iou(pbox.T, tbox[i], x1y1x2y2=False, CIoU=True)

The pbox is used to calculate the iou with tbox. Since tbox doesn't have c_x and c_y, pbox doesn't need c_x and c_y either.
